primitiv/c/shape.cc

```diff
@@ -5,239 +5,148 @@
-  SAFE_RETURN(primitiv_Shape_op_getitem(shape, i), status, 0);
+  return to_cpp_ptr(shape)->operator[](i);
 }
 
 void primitiv_Shape_dims(const primitiv_Shape *shape, uint32_t *array) {
-  std::vector<uint32_t> v = to_cc(shape)->dims();
+  std::vector<uint32_t> v = to_cpp_ptr(shape)->dims();
   std::copy(v.begin(), v.end(), array);
 }
-void safe_primitiv_Shape_dims(const primitiv_Shape *shape,
-                              uint32_t *array,
-                              primitiv_Status *status) {
-  SAFE_EXPR(primitiv_Shape_dims(shape, array), status);
-}
 
 uint32_t primitiv_Shape_depth(const primitiv_Shape *shape) {
-  return to_cc(shape)->depth();
-}
-uint32_t safe_primitiv_Shape_depth(const primitiv_Shape *shape,
-                                   primitiv_Status *status) {
-  SAFE_RETURN(primitiv_Shape_depth(shape), status, 0);
+  return to_cpp_ptr(shape)->depth();
 }
 
 uint32_t primitiv_Shape_batch(const primitiv_Shape *shape) {
-  return to_cc(shape)->batch();
-}
-uint32_t safe_primitiv_Shape_batch(const primitiv_Shape *shape,
-                                   primitiv_Status *status) {
-  SAFE_RETURN(primitiv_Shape_batch(shape), status, 0);
+  return to_cpp_ptr(shape)->batch();
 }
 
 uint32_t primitiv_Shape_volume(const primitiv_Shape *shape) {
-  return to_cc(shape)->volume();
-}
-uint32_t safe_primitiv_Shape_volume(const primitiv_Shape *shape,
-                                    primitiv_Status *status) {
-  SAFE_RETURN(primitiv_Shape_volume(shape), status, 0);
+  return to_cpp_ptr(shape)->volume();
 }
 
 uint32_t primitiv_Shape_lower_volume(const primitiv_Shape *shape,
                                      uint32_t dim) {
-  return to_cc(shape)->lower_volume(dim);
-}
-uint32_t safe_primitiv_Shape_lower_volume(const primitiv_Shape *shape,
-                                          uint32_t dim,
-                                          primitiv_Status *status) {
-  SAFE_RETURN(primitiv_Shape_lower_volume(shape, dim), status, 0);
+  return to_cpp_ptr(shape)->lower_volume(dim);
 }
 
 uint32_t primitiv_Shape_size(const primitiv_Shape *shape) {
-  return to_cc(shape)->size();
-}
-uint32_t safe_primitiv_Shape_size(const primitiv_Shape *shape,
-                                  primitiv_Status *status) {
-  SAFE_RETURN(primitiv_Shape_size(shape), status, 0);
+  return to_cpp_ptr(shape)->size();
 }
 
-char *primitiv_Shape_to_string(const primitiv_Shape *shape) {
-  std::string str = to_cc(shape)->to_string();
-  uint64_t len = str.length();
-  auto *c = new char[len + 1];
-  std::strncpy(c, str.c_str(), len);
-  return c;
-}
-char *safe_primitiv_Shape_to_string(const primitiv_Shape *shape,
-                                    primitiv_Status *status) {
-  SAFE_RETURN(primitiv_Shape_to_string(shape), status, nullptr);
+void primitiv_Shape_to_string(const primitiv_Shape *shape,
+                              char *string, size_t *length) {
+  std::string str = to_cpp_ptr(shape)->to_string();
+  *length = str.length();
```

Comment:
`str.length` returns the length of the sequence, except the NULL character at the end. The buffer for the resulting data should have at least `str.length() + 1` size.